arch/x86/pagetables.c

```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+
+    spin_lock(&mmap_lock);
+
+    for (; (cur % PAGE_SIZE_2M) && cur < end; cur += PAGE_SIZE)
+        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
+
+    if (likely(cur >= end))
+        goto done_4k;
+
+    for (; (cur % PAGE_SIZE_1G) && cur + PAGE_SIZE_2M <= end; cur += PAGE_SIZE_2M)
+        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
+
+    if (likely(cur >= end))
+        goto done_2m;
+
+    for (; cur + PAGE_SIZE_1G <= end; cur += PAGE_SIZE_1G)
+        MMAP_RANGE_1G(paddr_to_mfn(cur), flags, mmap_flags);
+
+    for (; cur + PAGE_SIZE_2M <= end; cur += PAGE_SIZE_2M)
+        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
+
+done_2m:
+    for (; cur < end; cur += PAGE_SIZE)
+        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
+
+done_4k:
+    ASSERT(paddr_to_mfn(cur) == paddr_to_mfn(end));
+    spin_unlock(&mmap_lock);
+
+    if (mmap_flags & MMAP_KERNEL_MAP)
+        return paddr_to_virt_map(paddr);
+    else if (mmap_flags & MMAP_KERNEL)
+        return paddr_to_virt_kern(paddr);
+    else if (mmap_flags & MMAP_USER)
+        return paddr_to_virt_user(paddr);
+    else if (mmap_flags & MMAP_IDENT)
+        return paddr_to_virt(paddr);
+    return NULL;
+}
+
+void munmap_range(void *va, size_t size, mmap_flags_t mmap_flags) {
+    paddr_t start = virt_to_paddr(va);
+    paddr_t end = start + size;
+
+    /* Round up to the next page boundary unless it is page aligned */
+    end = (end + PAGE_SIZE - 1) & PAGE_MASK;
+
+    spin_lock(&mmap_lock);
+
+    for (paddr_t cur = start; cur < end; cur += PAGE_SIZE) {
+        if (likely(mmap_flags & MMAP_KERNEL))
+            vunmap_kern(paddr_to_virt_kern(cur));
+        if (mmap_flags & MMAP_IDENT)
+            vunmap_kern(paddr_to_virt(cur));
+        if (mmap_flags & MMAP_USER)
+            vunmap_user(paddr_to_virt_user(cur));
+        if (mmap_flags & MMAP_USER_KERNEL)
+            vunmap_kern(paddr_to_virt_user(cur));
+        if (mmap_flags & MMAP_KERNEL_MAP)
+            vunmap_kern(paddr_to_virt_map(cur));
+    }
+
+    spin_unlock(&mmap_lock);
+}
+
 void init_pagetables(void) {
     for_each_memory_range (r) {
+        mmap_flags_t flags = MMAP_NONE;
+
         switch (r->base) {
         case VIRT_IDENT_BASE:
-            for (mfn_t mfn = virt_to_mfn(r->start); mfn < virt_to_mfn(r->end); mfn++)
-                vmap_4k(mfn_to_virt(mfn), mfn, r->flags);
+            flags = MMAP_IDENT;
             break;
         case VIRT_KERNEL_BASE:
-            for (mfn_t mfn = virt_to_mfn(r->start); mfn < virt_to_mfn(r->end); mfn++)
-                kmap_4k(mfn, r->flags);
+            flags = MMAP_KERNEL;
             break;
         case VIRT_USER_BASE:
-            for (mfn_t mfn = virt_to_mfn(r->start); mfn < virt_to_mfn(r->end); mfn++) {
-                void *va = mfn_to_virt_user(mfn);
-
-                vmap_4k(va, mfn, r->flags);
-                vmap_user_4k(va, mfn, r->flags);
-            }
+            flags = MMAP_USER | MMAP_USER_KERNEL;
```

Comment:
Oh, a combination of flags is actually a valid use case? If so, either the 'if (flag & FOO) ...; if (flag & BAR) ...;` pattern is broken or the if-else-if-else-if one is.

Reply:
Yes it is. We want to be able to map the `mfn` to either multiple memory regions or kernel and user pagetable or a combination thereof. I think the confusing part is what the `mmap_range()` returns. Maybe the previous comment above clarifies it a bit?